### ISSUE TITLE
input.habor to output.icecast in most efficient way?

### ISSUE THREAD

**Issue by S54B32** [savonet/liquidsoap#300]
What is the most efficient way for Liquidsoap to harbor a high number (40+) of streams and output in the same bitrate? In effect the Icecast relay feature...

I have been experimenting using a simple Liquidsoap script which basically has a harbor, takes in an MP3 and outputs it in MP3 to Icecast in the same bitrate/format.

What I've seen when running, say, 40 of these scripts concurrently on one machine that the CPU usage is very high, compared to say Icecast receiving 40 streams.

I'm wondering - am I going about this in the wrong way? I am guessing this method of input.harbor > output.icecast is still transcoding the audio?

Thanks

**Comment by StudentSA** [comment]
I am certainly no expert, and stand to be corrected. I believe Liquidsoap always performs encoding for output.icecast. i.e. you must specify some encoder (%mp3,%shine etc.).

I thought of a workaround involving specifying a external encoder that does nothing:
%external(channels=2,samplerate=44100,process="cat -")
However I could not get this to work for some reason.

Eventually to save cpu I just used my normal bitrates etc but set internal_quality to 9 which is the lowest cpu intensive:
%mp3(bitrate=32, samplerate=44100, mono=true, internal_quality=9)

hope that helps somewhat.

**Comment by smimram** [comment]
Just to clarify a bit: unfortunately, the current architecture of Liquidsoap makes it impossible to handle internally compressed data. This means that yes, Liquidsoap transcodes in your case, and there is now way to disable that :(

**Comment by S54B32** [comment]
Thanks both for your input - I suspected there wasn't a way of simply 'shifting' the stream without transcoding it.

I will try using the internal_quality parameter.

I have also tweaked the resampling quality which I've seen cut CPU usage in half from the default value when changing it to linear.

set("audio.converter.samplerate.libsamplerate.quality","zero_order")


Perhaps this in conjunction with the internal_quality parameter will make CPU usage more comparable without too much loss of quality.

Thanks again
